internal/fxlog/fxlog.go

```diff
@@ -21,68 +21,121 @@
-	Printf(string, ...interface{})
+// Level is the level of logging used by Logger.
+type Level int
+
+const (
+	InfoLevel = iota
+	ErrorLevel
+)
+
+// Entry is an entry
+type Entry struct {
+	Level   Level
+	Message string
+	Fields  []Field
+	Stack   string
 }
 
-// New returns a new Logger backed by the standard library's log package.
-func New() *Logger {
-	return &Logger{log.New(os.Stderr, "", log.LstdFlags)}
+func (e Entry) WithStack(stack string) Entry {
+	e.Stack = stack
+
+	return e
 }
 
-// A Logger writes output to standard error.
-type Logger struct {
-	Printer
+func (e Entry) Write(logger Logger) {
+	logger.Log(e)
 }
 
-// Printf logs a formatted Fx line.
-func (l *Logger) Printf(format string, v ...interface{}) {
-	l.Printer.Printf(prepend(format), v...)
+type Field struct {
+	Key   string
+	Value interface{}
 }
 
-// PrintProvide logs a type provided into the dig.Container.
-func (l *Logger) PrintProvide(t interface{}) {
-	for _, rtype := range fxreflect.ReturnTypes(t) {
-		l.Printf("PROVIDE\t%s <= %s", rtype, fxreflect.FuncName(t))
+type Logger interface {
+	Log(entry Entry)
+	PrintProvide(interface{})
+	PrintSupply(interface{})
+}
+
+var _ Logger = (*zapLogger)(nil)
+
+type zapLogger struct {
+	logger *zap.Logger
+}
+
+func encodeFields(fields []Field) []zap.Field {
+	var fs []zap.Field
+	for _, field := range fields {
+		fs = append(fs, zap.Any(field.Key, field.Value))
 	}
+
+	return fs
 }
 
-// PrintSupply logs a type supplied directly into the dig.Container
-// by the given constructor function.
-func (l *Logger) PrintSupply(constructor interface{}) {
-	for _, rtype := range fxreflect.ReturnTypes(constructor) {
-		l.Printf("SUPPLY\t%s", rtype)
+func (l *zapLogger) Log(entry Entry) {
+	switch entry.Level {
+	case InfoLevel:
+		l.logger.Info(entry.Message, encodeFields(entry.Fields)...)
+	case ErrorLevel:
+		l.logger.Error(entry.Message, encodeFields(entry.Fields)...)
 	}
 }
 
-// PrintSignal logs an os.Signal.
-func (l *Logger) PrintSignal(signal os.Signal) {
-	l.Printf(strings.ToUpper(signal.String()))
+func (l *zapLogger) PrintProvide(t interface{}) {
+	for _, rtype := range fxreflect.ReturnTypes(t) {
+		Info("providing",
+			Field{Key: "return value", Value: rtype},
+			Field{Key: "constructor", Value: fxreflect.FuncName(t)},
+		).Write(l)
+	}
 }
 
-// Panic logs an Fx line then panics.
-func (l *Logger) Panic(err error) {
-	l.Printer.Printf(prepend(err.Error()))
-	panic(err)
+func (l *zapLogger) PrintSupply(t interface{}) {
+	for _, rtype := range fxreflect.ReturnTypes(t) {
+		Info("supplying",
+			Field{Key: "constructor", Value: rtype},
+		).Write(l)
+	}
 }
 
-// Fatalf logs an Fx line then fatals.
-func (l *Logger) Fatalf(format string, v ...interface{}) {
-	l.Printer.Printf(prepend(format), v...)
-	_exit()
+// DefaultLogger constructs a Logger out of io.Writer.
+func DefaultLogger(ws zapcore.WriteSyncer) Logger {
+	zcore := zapcore.NewCore(
+		zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()),
+		zapcore.Lock(ws),
```

Comment:
We shouldn't add a mutex around the write syncer.

This is for when we know that the underlying implementation isn't thread-safe,
which isn't something we can assume. We can let the caller decide if they want
to add a locking wrapper around their write syncer.